### PR TITLE
Adding fetch implementation of request api, issue #139

### PR DESCRIPTION
**Pull request by rorticus**
Added `request/fetch` for performing requests based on the [`fetch` API](https://fetch.spec.whatwg.org/).

### CODE REVIEW

File: src/request/fetch.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	let request = new Request(requestUrl, fetchRequestOptions);
+
+	return new Task<RequestResponse<T>>((resolve, reject) => {
+		let timeout: Handle;
+
+		fetch(request).then((fetchResponse: any) => {
+			timeout && timeout.destroy();
+
+			let { responseType = '' } = options;
+			let body: any = null;
```
Comment (agubler) [inline src/request/fetch.ts]: can you type this as a `Promise<any>` (or not `any` if you know the body type) and then you wouldn't need the `Promise.resolve('body').then...` just simply `body.then...`

File: src/request/fetch.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+						return parser.parseFromString(asText, 'text/xml');
+					});
+					break;
+
+				default:
+					body = fetchResponse.text();
+					break;
+			}
+
+			let responseHeaders: any = {};
```
Comment (agubler) [inline src/request/fetch.ts]: any reason not `const`?

File: src/request/fetch.ts (hunk 1 above)
Comment (rorticus) [reply src/request/fetch.ts]: Yes! Good point. I guess I had typed that before I realized it was going to have to be a Promise, and then forgot to change it :/

File: src/request/fetch.ts (hunk 2 above)
Comment (agubler) [reply src/request/fetch.ts]: Also do you know the type of `headers`, presumably something like maybe

```ts
interface Headers {
    [index string]: string;
}
```

File: src/request/fetch.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+					body = fetchResponse.arrayBuffer();
+					break;
+
+				case 'blob':
+					body = fetchResponse.blob();
+					break;
+
+				case 'xml':
+					body = fetchResponse.text().then((asText: string) => {
+						let parser = new DOMParser();
```
Comment (agubler) [inline src/request/fetch.ts]: `const`?

File: src/request/fetch.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+	let request = new Request(requestUrl, fetchRequestOptions);
+
+	return new Task<RequestResponse<T>>((resolve, reject) => {
+		let timeout: Handle;
+
+		fetch(request).then((fetchResponse: any) => {
+			timeout && timeout.destroy();
+
+			let { responseType = '' } = options;
```
Comment (agubler) [inline src/request/fetch.ts]: `const`?

File: src/request/fetch.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+export interface FetchRequestOptions extends RequestOptions {
+}
+
+export default function fetchRequest<T>(url: string, options: FetchRequestOptions = {}): ResponsePromise<T> {
+	const fetchRequestOptions: any = {};
+	const fetchRequestHeaders: Headers = new Headers();
+	const requestUrl = generateRequestUrl(url, options);
+
+	if ((!options.user || !options.password) && options.auth) {
+		let auth = options.auth.split(':');
```
Comment (agubler) [inline src/request/fetch.ts]: `const`?

File: src/request/fetch.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+	}
+
+	fetchRequestOptions.method = options.method;
+
+	if (options.headers) {
+		const headers = options.headers;
+		let hasContentTypeHeader = false;
+		let hasRequestedWithHeader = false;
+
+		for (let header in headers) {
```
Comment (agubler) [inline src/request/fetch.ts]: sorry about all the `const` suggestions...

File: src/request/fetch.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+					break;
+			}
+
+			let responseHeaders: any = {};
+
+			forOf(fetchResponse.headers.keys(), (key: string) => {
+				responseHeaders[ key.toLowerCase() ] = fetchResponse.headers.get(key);
+			});
+
+			Promise.resolve(body).then((body: any) => {
```
Comment (agubler) [inline src/request/fetch.ts]: call this `data` and you can just use `data` in the resolved object and drop the `: body`

File: src/request/fetch.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+						return responseHeaders[ name.toLowerCase() ];
+					}
+				});
+			}, reject);
+		}, reject);
+
+		if (options.timeout > 0 && options.timeout !== Infinity) {
+			timeout = (function (): Handle {
+				const timer = setTimeout(function (): void {
+					const error = new RequestTimeoutError('Request timed out after ' + options.timeout + 'ms');
```
Comment (agubler) [inline src/request/fetch.ts]: how about using template literals 

```ts
new RequestTimeoutError(`Request timed out after ${options.timeout}ms`);
```

File: src/request/fetch.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+					reject(error);
+				}, options.timeout);
+
+				return createHandle(function (): void {
```
Comment (agubler) [inline src/request/fetch.ts]: can do if you so desire,

```ts
				return createHandle((): void => {
					clearTimeout(timer);
				});
```

or perhaps even

```ts
			timeout = ((): Handle => {
				const timer = setTimeout((): void => {
					const error = new RequestTimeoutError('Request timed out after ' + options.timeout + 'ms');
					reject(error);
				}, options.timeout);

				return createHandle(clearTimeout.bind(null, timer));
			})();
```

**Review by agubler — CHANGES_REQUESTED** [review]
Just a couple minor comments/questions.

File: src/request/fetch.ts (hunk 5 above)
Comment (rorticus) [reply src/request/fetch.ts]: Done!

File: src/request/fetch.ts (hunk 6 above)
Comment (rorticus) [reply src/request/fetch.ts]: Done!

File: src/request/fetch.ts (hunk 4 above)
Comment (rorticus) [reply src/request/fetch.ts]: Done!

File: src/request/fetch.ts (hunk 3 above)
Comment (rorticus) [reply src/request/fetch.ts]: Done!

File: src/request/fetch.ts (hunk 2 above)
Comment (rorticus) [reply src/request/fetch.ts]: Done!

File: src/request/fetch.ts (hunk 8 above)
Comment (rorticus) [reply src/request/fetch.ts]: Done!

File: src/request/fetch.ts (hunk 9 above)
Comment (rorticus) [reply src/request/fetch.ts]: Done!

File: src/request/fetch.ts [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+					break;
+
+				default:
+					body = fetchResponse.text();
+					break;
+			}
+
+			const responseHeaders: ResponseHeaders = {};
+
+			forOf(fetchResponse.headers.keys(), (key: string) => {
```
Comment (agubler) [inline src/request/fetch.ts]: another mainly styling suggestion (I think that works 😉 )

```ts
			forOf(fetchResponse.headers.entries(), ([key, value]) => {
				responseHeaders[ key.toLowerCase() ] = value;
			});
```

File: src/request/fetch.ts [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+					break;
+
+				default:
+					body = fetchResponse.text();
+					break;
+			}
+
+			const responseHeaders: ResponseHeaders = {};
+
+			forOf(fetchResponse.headers.enrties(), ([key, value]) => {
```
Comment (agubler) [inline src/request/fetch.ts]: oops typo.

Comment (rorticus) [reply src/request/fetch.ts]: Ugh. how did that build? ....

Comment (agubler) [reply src/request/fetch.ts]: I really do not know!

File: src/request/fetch.ts [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+				case 'arraybuffer':
+					body = fetchResponse.arrayBuffer();
+					break;
+
+				case 'blob':
+					body = fetchResponse.blob();
+					break;
+
+				case 'xml':
+					body = fetchResponse.text().then((asText: string) => {
```
Comment (agubler) [inline src/request/fetch.ts]: I believe `.then` should be on a new line as per the style guide

Comment (rorticus) [reply src/request/fetch.ts]: +1

**Review by agubler — APPROVED** [review]
Approved from me! Someone else might want eyes on before the big green light.

**Comment by agubler** [comment]
Not sure if anyone else will get the opportunity to review this 🎱

**Comment by dylans** [comment]
A few stray thoughts that are probably worth addressing mostly later as we decide how useful the fetch API might be:

* the fetch API doesn't seem to have a convenience method for json, so the user must explicitly parse json? Is that how our request implementation works also?
* the fetch API seems to implement ReadableStream... I assume we're planning to ignore that for now given the size of the streams shim?
* are we ignoring the type Document (I see we have arraybuffer and blob), or does that get parsed the same way as plain text?
* the fetch API has a number of scenarios around cors, requestType, requestDestination, etc. Do we plan to/need to support these features now or in the future?

**Comment by rorticus** [comment]
> the fetch API doesn't seem to have a convenience method for json, so the user must explicitly parse json? Is that how our request implementation works also?

We have [a response filter](https://github.com/dojo/core/blob/master/src/request.ts#L198-L207) that runs at a "higher level" that will turn the json type into an actual JSON object.

> the fetch API seems to implement ReadableStream... I assume we're planning to ignore that for now given the size of the streams shim?

I'm OK with ignoring that :)  If we do decide to implement it, we should at least figure out how we're going to pull in `streams` without the dependency in the node request.

> are we ignoring the type Document (I see we have arraybuffer and blob), or does that get parsed the same way as plain text?

It'll get parsed as plain text.

> the fetch API has a number of scenarios around cors, requestType, requestDestination, etc. Do we plan to/need to support these features now or in the future?

That is a great question :) I've ignored them for now because we didn't have anything similar in the xhr/node requests, but adding those options might be a compelling reason why someone would _want_ to use the fetch API over the alternatives.

**Comment by rorticus** [comment]
Dylan's comments have been split into https://github.com/dojo/core/issues/255 and https://github.com/dojo/core/issues/254 to get this landed.

**Pull request merged**